nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+
+// User update routes
+usersRouter.get("/users/:id/update", usersController.user_update_get);
+usersRouter.post("/users/:id/update", usersController.user_update_post);
+
+module.exports = usersRouter;
+```
+
+You'll notice there's a respective `GET` and `POST` route.
+
+- `GET` routes typically display forms or confirmation pages.
+- `POST` routes handle form submissions and data processing.
+
+You may also notice the `:id` parameter in the route. This lets us pass a value to those routes that we can use to tell the controller which `User` object we're referencing.
+
+And in our form, the action would look something like this:
+
+```ejs
+<!-- Example using EJS with POST to submit an update to our Express server. -->
+<form action="/users/<%= user.userId %>/update" method="POST"></form>
+```
+
+`/users/:id/update` is an *endpoint* we've created on our Express server.
+
+### Handling errors
+
+What happens if we don't give our route a valid `:id` to delete or update? We need a way to tell our users an error occurred.
+
+```javascript
+// Handles POST requests for the /users/:id/delete route.
+// Status code 400 can indicate malformed request syntax, or a missing :id in our case.
+exports.user_delete_post = asyncHandler(async (req, res, next) => {
+  if (!req.params.id) {
+    return res.status(400).send("No user id provided.");
+  }
+});
+```
+
+By responding with a status code and a message, we give the client a way to both handle the error code, and a way to display an appropriate error message. We could also render a new or updated view which we'll see below.
+
+### Putting it together
+
+Let's create a quick Express application to demonstrate how clients send data to servers with forms.
+
+Set up a new Express app with EJS templating:
+
+```bash
+npm init -y
+npm install express ejs express-async-handler express-validator
+```
+
+Create folders for `routes`, `views`, `controllers`, and an `app.js` file:
+
+```javascript
+// app.js
+const express = require("express");
+const app = express();
+const usersRouter = require("./routes/usersRouter");
+
+app.set("view engine", "ejs");
+app.use(express.urlencoded({ extended: true }));
+app.use("/users", usersRouter);
+app.get("/", (req, res) => res.render("index", { title: "Userbase" }));
+
+const PORT = process.env.PORT || 3000;
+app.listen(PORT, () => console.log(`Express app listening on port ${PORT}!`));
+```
+
+Then create a new router called `usersRouter.js` in the routes folder:
+
+```javascript
+// routes/usersRouter.js
+const { Router } = require("express");
+const usersController = require("../controllers/usersController");
+const usersRouter = Router();
+
+usersRouter.get("/", usersController.users_list);
+usersRouter.post("/create", usersController.users_create_post);
+
+module.exports = usersRouter;
+```
+
+Next we'll create two views:
+
+- `index.ejs` will display our main form.
+- `users.ejs` will list all the users we've created.
```

Comment:
I'd rather have this switched. To me it makes more sense for the index page to display data and a link to the form page. What do you think?

Reply:
Yeah makes sense to me. I think the logic came from an earlier iteration I wrote, but I'm good to switch it 👍